contracts/Witch.sol

```diff
@@ -69,138 +75,205 @@ contract Witch is AccessControl() {
-    function auction(bytes12 vaultId)
-        external
-    {
-        require (auctions[vaultId].start == 0, "Vault already under auction");
-        require (cauldron.level(vaultId) < 0, "Not undercollateralized");
+    /// @dev Put an undercollateralized vault up for liquidation
+    /// @param vaultId Id of vault to liquidate
+    function auction(bytes12 vaultId) external {
+        require(auctions[vaultId].start == 0, "Vault already under auction");
+        require(cauldron.level(vaultId) < 0, "Not undercollateralized");
 
         DataTypes.Vault memory vault_ = cauldron.vaults(vaultId);
         DataTypes.Balances memory balances_ = cauldron.balances(vaultId);
         Limits memory limits_ = limits[vault_.ilkId];
         limits_.sum += balances_.ink;
-        require (limits_.sum <= limits_.line * (10 ** limits_.dec), "Collateral limit reached");
+        require(limits_.sum <= limits_.line * (10**limits_.dec), "Collateral limit reached");
 
         limits[vault_.ilkId] = limits_;
-        auctions[vaultId] = Auction({
-            owner: vault_.owner,
-            start: block.timestamp.u32()
-        });
+        auctions[vaultId] = Auction({owner: vault_.owner, start: block.timestamp.u32()});
         cauldron.give(vaultId, address(this));
         emit Auctioned(vaultId, block.timestamp.u32());
     }
 
+    /// @dev Private fn called by buy() and payAll()
+    /// @param ilkId Id of asset used for collateral
+    /// @param auctionStart Block timestamp when auction was started
+    /// @param duration Time that auctions take to go to minimal
+    /// @param p Initial Offer. Proportion of collateral sold at auction start(1e18 = 100%)
+    /// @param artIn Portion of debt being bought (in terms of base)
+    /// @param totalArt Total debt
+    /// @param totalInk Total collateral
+    /// @param min Minimum amount of collateral acceptable by buyer
+    /// @return inkOut Amount of collateral
+    function _inkOut(
+        bytes6 ilkId,
+        uint32 auctionStart,
+        uint32 duration,
+        uint64 p,
+        uint256 artIn,
+        uint128 totalArt,
+        uint128 totalInk,
+        uint128 min
+    ) private returns (uint256 inkOut) {
+        Limits memory limits_ = limits[ilkId];
+
+        // If the world has not turned to ashes and darkness, auctions will malfunction on
+        // the 7th of February 2106, at 06:28:16 GMT
+        // TODO: Replace this contract before then 😰
+        // UPDATE: Added reminder to Google calendar ✅
+        uint256 elapsed;
+        unchecked {
+            elapsed = uint32(block.timestamp) - auctionStart;
+        }
+
+        uint256 t = elapsed > duration ? 1e18 : elapsed.wdivup(duration);
+
+        //          (      a         )
+        // inkOut = (artIn / totalArt) * totalInk * (p + (1 - p) * t)
+        {
+            uint256 a = uint256(artIn).wdivup(totalArt);
+            inkOut = a.wmul(totalInk).wmulup(uint256(p) + uint256(1e18 - p).wmulup(t));
```

Comment:
Do you need to round up in any arithmetic operation in this function?

Reply:
I did my best to stick with the rounding from the original function.  Pretty much everything was a straight port accept for the initial term.  Re: the initial term, the old version was:
```
totalnk.divup(totalArk).mulup(<next term>...
```

so I interpret the new fn as:
```
(artIn.divup(totalArt)).mulup(<next term>...
```

WDYT?